### PR TITLE
feat: Port ScrollView

### LINKED ISSUE #14370: Add support for `ScrollView`
### What would you like to be added

_No response_

### Why is this needed

_No response_

### For which platform

WebAssembly, Android, iOS, macOS (AppKit), Mac Catalyst, Skia (WPF), Skia (GTK on Linux/macOS/Windows), Skia (Linux Framebuffer)

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #14370

## PR Type

What kind of change does this PR introduce?

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at d765284</samp>

This pull request adds the `ScrollPresenter` class and its related classes, interfaces, and enums to the Uno.UI project. The `ScrollPresenter` class is a UI element that can scroll and zoom its content, and it supports various scrolling modes, snap points, and input kinds. The pull request also adds the UI automation support and the gamepad focus navigation helper for the `ScrollPresenter` class. Additionally, it modifies the signature of the `IsAncestor` method in the `SharedHelpers` class, and adds the license header and the using directives to the `FocusHelper.cs` file.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/IScrollPresenter.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+	event TypedEventHandler<ScrollPresenter, ScrollingZoomCompletedEventArgs> ZoomCompleted;
+
+	int ScrollTo(double horizontalOffset, double verticalOffset);
+
+	int ScrollTo(double horizontalOffset, double verticalOffset, ScrollingScrollOptions options);
+
+	int ScrollBy(double horizontalOffsetDelta, double verticalOffsetDelta);
+
+	int ScrollBy(double horizontalOffsetDelta, double verticalOffsetDelta, ScrollingScrollOptions options);
```
Comment (dr1rrb) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/IScrollPresenter.cs:107]: Is this interface coming from WinUI ? If no, in order to make it easier to implement, why not having only one `Scroll` (or at least one `ScolrBy` and one `ScrollTo`) instead? (Same for `Zoom`)

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/ScrollingAnchorRequestedEventArgs.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+using System;
+using System.Collections.Generic;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls.Primitives;
+
+namespace Microsoft.UI.Xaml.Controls;
+
+public sealed partial class ScrollingAnchorRequestedEventArgs
+{
+	public ScrollingAnchorRequestedEventArgs(ScrollPresenter scrollPresenter)
```
Comment (dr1rrb) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/ScrollingAnchorRequestedEventArgs.cs:13]: This needs to be `internal`

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/SnapPoint.cs [hunk 3]
```diff
@@ -0,0 +1,1920 @@
...
+	// HOWEVER, the API seen in a real WinUI app says that the parameter is mandatory.
+	public ScrollSnapPoint(
+		double snapPointValue,
+		ScrollSnapPointsAlignment alignment)
+	{
+		Value = snapPointValue;
+		Alignment = alignment;
+	}
+
+#if ApplicableRangeType // UNO TODO:
```
Comment (dr1rrb) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/SnapPoint.cs:181]: Is this from WInUI? Compiler variables are usually UPPER_SNAKE_CASED

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/SnapPoint.cs [hunk 4]
```diff
@@ -0,0 +1,1920 @@
...
+using Microsoft.UI.Composition;
+using Microsoft.UI.Composition.Interactions;
+using Uno.UI.Helpers.WinUI;
+using Windows.UI;
+
+using static Microsoft/* UWP don't rename */.UI.Xaml.Controls._Tracing;
+
+namespace Microsoft.UI.Xaml.Controls.Primitives;
+
+public partial class SnapPointBase
```
Comment (dr1rrb) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/SnapPoint.cs:14]: I think that you need to add an `internal` ctor on this

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollView/ScrollView.cs [hunk 5]
```diff
@@ -0,0 +1,2692 @@
...
+#endif
+public partial class ScrollView : Control, IScrollView
+{
+	// Change to 'true' to turn on debugging outputs in Output window
+	//bool ScrollViewTrace::s_IsDebugOutputEnabled{ false };
+	//bool ScrollViewTrace::s_IsVerboseDebugOutputEnabled{ false };
+
+	private const int s_noOpCorrelationId = -1;
+
+	public ScrollView()
```
Comment (dr1rrb) [inline src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollView/ScrollView.cs:41]: Didn't validated if already done, but this class has to [`RegisterAsScrollPort`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.registerasscrollport?view=winrt-22621), update the [`ScrollOffsets `](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/UIElement.cs#L76)  and then call [`InvalidateViewport`](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.frameworkelement.invalidateviewport?view=winrt-22621) in order to raise the `EffetiveViewportChange` event properly (and be usable with the `ItemRepeater`).

**Review by dr1rrb — COMMENTED** [review]
I'm surprised that adding the `ScrollView` does not have any impact in the `UIElement.TransformToVisual` nor the `VisualTreeHelper.HitTtest`. Is this PR only one of a larger work?

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/IScrollPresenter.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/IScrollPresenter.cs:107]: It's from WinUI source code

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/ScrollingAnchorRequestedEventArgs.cs (hunk 2 above)
Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/ScrollingAnchorRequestedEventArgs.cs:13]:
```suggestion
	internal ScrollingAnchorRequestedEventArgs(ScrollPresenter scrollPresenter)
```

File: src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/SnapPoint.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Uno.UI/Microsoft/UI/Xaml/Controls/ScrollPresenter/SnapPoint.cs:181]: @dr1rrb Yes, from WinUI, and it was quite unclear to me if it should be true or false in our context. So left it as a TODO

**Review by dr1rrb — APPROVED** [review]

**Pull request merged**